lmdeploy/pytorch/engine/guided_process.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2024- the Outlines developers
+# This file is adapted from
+# https://github.com/outlines-dev/outlines/blob/main/outlines/serve/vllm.py
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+import copy
+import math
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+from functools import lru_cache
+from typing import DefaultDict, Dict, List, Union
+
+import torch
+from outlines.fsm.fsm import RegexFSM
+from outlines.fsm.json_schema import build_regex_from_schema
+from pydantic import BaseModel
+from transformers import PreTrainedTokenizerBase
+
+
+class RegexLogitsProcessor:
+
+    def __init__(self, regex_string: str, tokenizer):
+        """Compile the FSM that drives the regex-structured generation.
+
+        Args:
+            regex_string: A string that represents a regular expression
+            tokenizer: The model's tokenizer
+        """
+        tokenizer = self.adapt_tokenizer(copy.deepcopy(tokenizer))
+        fsm = RegexFSM(regex_string, tokenizer)
+        self.fsm = fsm
+
+    def init_state(self):
+        """Initialize the FSM states."""
+        self.fsm_state: DefaultDict[int, int] = defaultdict(int)
```

Comment:
What if two sequences in batch use the same guide, they will share the same processor (because of lru).
the second sequence might clear the fsm_state of the first sequence.

Reply:
And the state won't be cleared when the sequence is ended.
Can we store the fsm_state in sequence?

Reply:
`fsm_state` should be independent of `sequence` since some requests may require guided decoding while some are not. Or else, we must maintain the reference count of `fsm_state`.

Reply:
Any request with `len(input_ids)` will clear the fsm_state of other requests with the same regex_string, I don't think that is expected.
And if we send `stop` or `end` request, will the state be cleared?